File: samples/csharp_dotnetcore/44.prompt-users-for-input/Bots/CustomPromptBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using Microsoft.Recognizers.Text;
using Microsoft.Recognizers.Text.DateTime;
using Microsoft.Recognizers.Text.Number;

namespace Microsoft.BotBuilderSamples
{
    // This IBot implementation can run any type of Dialog. The use of type parameterization is to allows multiple different bots
    // to be run at different endpoints within the same project. This can be achieved by defining distinct Controller types
    // each with dependency on distinct IBot types, this way ASP Dependency Injection can glue everything together without ambiguity.
    public class CustomPromptBot : ActivityHandler 
    {
        private readonly BotState _userState;
        private readonly BotState _conversationState;
        
        public CustomPromptBot(ConversationState conversationState, UserState userState)
        {
            _conversationState = conversationState;
            _userState = userState;
        }

        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
           
            var conversationStateAccessors = _conversationState.CreateProperty<ConversationFlow>(nameof(ConversationFlow));
            var flow = await conversationStateAccessors.GetAsync(turnContext, () => new ConversationFlow(), cancellationToken);

            var userStateAccessors = _userState.CreateProperty<UserProfile>(nameof(UserProfile));
            var profile = await userStateAccessors.GetAsync(turnContext, () => new UserProfile(), cancellationToken);

            await FillOutUserProfileAsync(flow, profile, turnContext);

            // Save changes.
            await _conversationState.SaveChangesAsync(turnContext, cancellationToken: cancellationToken);
            await _userState.SaveChangesAsync(turnContext, cancellationToken: cancellationToken);
        }

        private static async Task FillOutUserProfileAsync(ConversationFlow flow, UserProfile profile, ITurnContext turnContext)
        {
            string input = turnContext.Activity.Text?.Trim();
            string message;

            switch (flow.LastQuestionAsked)
            {
                case ConversationFlow.Question.None:
                    await turnContext.SendActivityAsync("Let's get started. What is your name?");
                    flow.LastQuestionAsked = ConversationFlow.Question.Name;
                    break;
                case ConversationFlow.Question.Name:
                    if (ValidateName(input, out string name, out message))
                    {
                        profile.Name = name;
                        await turnContext.SendActivityAsync($"Hi {profile.Name}.");
                        await turnContext.SendActivityAsync("How old are you?");
                        flow.LastQuestionAsked = ConversationFlow.Question.Age;
                        break;
                    }
                    else
                    {
                        await turnContext.SendActivityAsync(message ?? "I'm sorry, I didn't understand that.");
                        break;
                    }
                case ConversationFlow.Question.Age:
                    if (ValidateAge(input, out int age, out message))
                    {
                        profile.Age = age;
                        await turnContext.SendActivityAsync($"I have your age as {profile.Age}.");
                        await turnContext.SendActivityAsync("When is your flight?");
                        flow.LastQuestionAsked = ConversationFlow.Question.Date;
                        break;
                    }
                    else
                    {
                        await turnContext.SendActivityAsync(message ?? "I'm sorry, I didn't understand that.");
                        break;
                    }

                case ConversationFlow.Question.Date:
                    if (ValidateDate(input, out string date, out message))
                    {
                        profile.Date = date;
                        await turnContext.SendActivityAsync($"Your cab ride to the airport is scheduled for {profile.Date}.");
                        await turnContext.SendActivityAsync($"Thanks for completing the booking {profile.Name}.");
                        await turnContext.SendActivityAsync($"Type anything to run the bot again.");
                        flow.LastQuestionAsked = ConversationFlow.Question.None;
                        profile = new UserProfile();
                        break;
                    }
                    else
                    {
                        await turnContext.SendActivityAsync(message ?? "I'm sorry, I didn't understand that.");
                        break;
                    }
            }
        }

        private static bool ValidateName(string input, out string name, out string message)
        {
            name = null;
            message = null;

            if (string.IsNullOrWhiteSpace(input))
            {
                message = "Please enter a name that contains at least one character.";
            }
            else
            {
                name = input.Trim();
            }

            return message is null;
        }

        private static bool ValidateAge(string input, out int age, out string message)
        {
            age = 0;
            message = null;

            // Try to recognize the input as a number. This works for responses such as "twelve" as well as "12".
            try
            {
                // Attempt to convert the Recognizer result to an integer. This works for "a dozen", "twelve", "12", and so on.
                // The recognizer returns a list of potential recognition results, if any.

                var results = NumberRecognizer.RecognizeNumber(input, Culture.English);

                foreach (var result in results)
                {
                    // The result resolution is a dictionary, where the "value" entry contains the processed string.
                    if (result.Resolution.TryGetValue("value", out object value))
                    {
                        age = Convert.ToInt32(value);
                        if (age >= 18 && age <= 120)
                        {
                            return true;
                        }
                    }
                }

                message = "Please enter an age between 18 and 120.";
            }
            catch
            {
                message = "I'm sorry, I could not interpret that as an age. Please enter an age between 18 and 120.";
            }

            return message is null;
        }

        private static bool ValidateDate(string input, out string date, out string message)
        {
            date = null;
            message = null;

            // Try to recognize the input as a date-time. This works for responses such as "11/14/2018", "9pm", "tomorrow", "Sunday at 5pm", and so on.
            // The recognizer returns a list of potential recognition results, if any.
            try
            {
                var results = DateTimeRecognizer.RecognizeDateTime(input, Culture.English);

                // Check whether any of the recognized date-times are appropriate,
                // and if so, return the first appropriate date-time. We're checking for a value at least an hour in the future.
                var earliest = DateTime.Now.AddHours(1.0);

                foreach (var result in results)
                {
                    // The result resolution is a dictionary, where the "values" entry contains the processed input.
                    var resolutions = result.Resolution["values"] as List<Dictionary<string, string>>;

                    foreach (var resolution in resolutions)
                    {
                        // The processed input contains a "value" entry if it is a date-time value, or "start" and
                        // "end" entries if it is a date-time range.
                        if (resolution.TryGetValue("value", out string dateString)
                            || resolution.TryGetValue("start", out dateString))
                        {
                            if (DateTime.TryParse(dateString, out DateTime candidate)
                                && earliest < candidate)
                            {
                                date = candidate.ToShortDateString();
                                return true;
                            }
                        }
                    }
                }

                message = "I'm sorry, please enter a date at least an hour out.";
            }
            catch
            {
                message = "I'm sorry, I could not interpret that as an appropriate date. Please enter a date at least an hour out.";
            }

            return false;
        }
    }
}

File: samples/java_springboot/44.prompt-users-for-input/src/main/java/com/microsoft/bot/sample/promptusersforinput/CustomPromptBot.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.bot.sample.promptusersforinput;

import com.microsoft.bot.builder.ActivityHandler;
import com.microsoft.bot.builder.BotState;
import com.microsoft.bot.builder.ConversationState;
import com.microsoft.bot.builder.StatePropertyAccessor;
import com.microsoft.bot.dialogs.prompts.PromptCultureModels;
import com.microsoft.recognizers.text.ModelResult;
import com.microsoft.recognizers.text.datetime.DateTimeRecognizer;
import com.microsoft.recognizers.text.number.NumberRecognizer;
import com.microsoft.bot.builder.TurnContext;
import com.microsoft.bot.builder.UserState;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicReference;

import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.tuple.Triple;
import org.springframework.stereotype.Component;

/**
 * This Bot implementation can run any type of Dialog. The use of type
 * parameterization is to allows multiple different bots to be run at different
 * endpoints within the same project. This can be achieved by defining distinct
 * Controller types each with dependency on distinct Bot types.
 */
public class CustomPromptBot extends ActivityHandler {

    private final BotState userState;
    private final BotState conversationState;

    public CustomPromptBot(ConversationState conversationState, UserState userState) {
        this.conversationState = conversationState;
        this.userState = userState;
    }

    @Override
    protected CompletableFuture<Void> onMessageActivity(TurnContext turnContext) {

        StatePropertyAccessor<ConversationFlow> conversationStateAccessors =
            conversationState.createProperty("ConversationFlow");

        StatePropertyAccessor<UserProfile> userStateAccessors = userState.createProperty("UserProfile");
        return userStateAccessors.get(turnContext, () -> new UserProfile()).thenCompose(profile -> {
            return conversationStateAccessors.get(turnContext, () -> new ConversationFlow()).thenCompose(flow -> {
                return fillOutUserProfile(flow, profile, turnContext);
            });
        })
        .thenCompose(result -> conversationState.saveChanges(turnContext))
        .thenCompose(result -> userState.saveChanges(turnContext));
    }

    private static CompletableFuture<Void> fillOutUserProfile(ConversationFlow flow,
                                                              UserProfile profile,
                                                              TurnContext turnContext) {
        String input = "";
        if (StringUtils.isNotBlank(turnContext.getActivity().getText())) {
            input = turnContext.getActivity().getText().trim();
        }

        switch (flow.getLastQuestionAsked()) {
            case None:
                return turnContext.sendActivity("Let's get started. What is your name?", null, null)
                    .thenRun(() -> {flow.setLastQuestionAsked(ConversationFlow.Question.Name);});
            case Name:
                Triple<Boolean, String, String> nameValidationResult = validateName(input);
                if (nameValidationResult.getLeft()) {
                    profile.name = nameValidationResult.getMiddle();
                    return turnContext.sendActivity(String.format("Hi %s.", profile.name), null, null)
                        .thenCompose(result -> turnContext.sendActivity("How old are you?", null, null))
                        .thenRun(() -> { flow.setLastQuestionAsked(ConversationFlow.Question.Age); });
                } else {
                    if (StringUtils.isNotBlank(nameValidationResult.getRight())) {
                        return turnContext.sendActivity(nameValidationResult.getRight(), null, null)
                               .thenApply(result -> null);
                    } else {
                        return turnContext.sendActivity("I'm sorry, I didn't understand that.", null, null)
                               .thenApply(result -> null);
                    }
                }
            case Age:
                Triple<Boolean, Integer, String> ageValidationResult = ValidateAge(input);
                if (ageValidationResult.getLeft()) {
                    profile.age = ageValidationResult.getMiddle();
                    return turnContext.sendActivity(String.format("I have your age as %d.", profile.age), null, null)
                        .thenCompose(result -> turnContext.sendActivity("When is your flight?", null, null))
                        .thenRun(() -> { flow.setLastQuestionAsked(ConversationFlow.Question.Date); });
                } else {
                    if (StringUtils.isNotBlank(ageValidationResult.getRight())) {
                        return turnContext.sendActivity(ageValidationResult.getRight(), null, null)
                            .thenApply(result -> null);
                    } else {
                        return turnContext.sendActivity("I'm sorry, I didn't understand that.", null, null)
                            .thenApply(result -> null);
                    }
                }

            case Date:
                Triple<Boolean, String, String> dateValidationResult = ValidateDate(input);
                AtomicReference<UserProfile> profileReference = new AtomicReference<UserProfile>(profile);
                if (dateValidationResult.getLeft()) {
                    profile.date = dateValidationResult.getMiddle();
                    return turnContext.sendActivity(
                        String.format("Your cab ride to the airport is scheduled for %s.",
                                      profileReference.get().date))
                    .thenCompose(result -> turnContext.sendActivity(
                        String.format("Thanks for completing the booking %s.", profileReference.get().name)))
                    .thenCompose(result -> turnContext.sendActivity("Type anything to run the bot again."))
                    .thenRun(() -> {
                        flow.setLastQuestionAsked(ConversationFlow.Question.None);
                        profileReference.set(new UserProfile());
                    });
                } else {
                    if (StringUtils.isNotBlank(dateValidationResult.getRight())) {
                        return turnContext.sendActivity(dateValidationResult.getRight(), null, null)
                            .thenApply(result -> null);
                    } else {
                        return turnContext.sendActivity("I'm sorry, I didn't understand that.", null, null)
                            .thenApply(result -> null);                    }
                }
            default:
                return CompletableFuture.completedFuture(null);
        }
    }

    private static Triple<Boolean, String, String> validateName(String input) {
        String name = null;
        String message = null;

        if (StringUtils.isEmpty(input)) {
            message = "Please enter a name that contains at least one character.";
        } else {
            name = input.trim();
        }

        return Triple.of(StringUtils.isBlank(message), name, message);
    }

    private static Triple<Boolean, Integer, String> ValidateAge(String input) {
        int age = 0;
        String message = null;

        // Try to recognize the input as a number. This works for responses such as "twelve" as well as "12".
        try {
            // Attempt to convert the Recognizer result to an integer. This works for "a dozen", "twelve", "12", and so on.
            // The recognizer returns a list of potential recognition results, if any.

            List<ModelResult> results = NumberRecognizer.recognizeNumber(input, PromptCultureModels.ENGLISH_CULTURE);
            for (ModelResult result : results) {
                // The result resolution is a dictionary, where the "value" entry contains the processed String.
                Object value = result.resolution.get("value");
                if (value != null) {
                    age = Integer.parseInt((String) value);
                    if (age >= 18 && age <= 120) {
                        return Triple.of(true, age, "");
                    }
                }
            }

            message = "Please enter an age between 18 and 120.";
        }
        catch (Throwable th) {
            message = "I'm sorry, I could not interpret that as an age. Please enter an age between 18 and 120.";
        }

        return Triple.of(StringUtils.isBlank(message), age, message);
    }

    private static Triple<Boolean, String, String> ValidateDate(String input) {
        String date = null;
        String message = null;

        // Try to recognize the input as a date-time. This works for responses such as "11/14/2018", "9pm", "tomorrow", "Sunday at 5pm", and so on.
        // The recognizer returns a list of potential recognition results, if any.
        try {
            List<ModelResult> results = DateTimeRecognizer.recognizeDateTime(input, PromptCultureModels.ENGLISH_CULTURE);

            // Check whether any of the recognized date-times are appropriate,
            // and if so, return the first appropriate date-time. We're checking for a value at least an hour in the future.
            LocalDateTime earliest = LocalDateTime.now().plus(1, ChronoUnit.HOURS);

            for (ModelResult result : results) {
                // The result resolution is a dictionary, where the "values" entry contains the processed input.
                List<Map<String, Object>> resolutions = (List<Map<String, Object>>) result.resolution.get("values");

                 for (Map<String, Object> resolution : resolutions) {
                    // The processed input contains a "value" entry if it is a date-time value, or "start" and
                    // "end" entries if it is a date-time range.
                    String dateString = (String) resolution.get("value");
                    if (StringUtils.isBlank(dateString)) {
                        dateString = (String) resolution.get("start");
                    }
                    if (StringUtils.isNotBlank(dateString)){
                        DateTimeFormatter f = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
                        LocalDateTime candidate = LocalDateTime.from(f.parse(dateString));
                        if (earliest.isBefore(candidate)) {
                            DateTimeFormatter dateformat = DateTimeFormatter.ofPattern("MM-dd-yyyy");
                            date = candidate.format(dateformat);
                            return Triple.of(true, date, message);
                        }
                    }

                 }
            }

            message = "I'm sorry, please enter a date at least an hour out.";
        }
        catch (Throwable th) {
            message = "I'm sorry, I could not interpret that as an appropriate date. Please enter a date at least an hour out.";
        }

        return Triple.of(false, date, message);
    }
}
